pkg/cmd/experimental/syncgroups/ad/ldapinterface.go

```diff
@@ -0,0 +1,114 @@
+package ad
+
+import (
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/go-ldap/ldap"
+
+	"github.com/openshift/origin/pkg/auth/ldaputil"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+)
+
+// NewLDAPInterface builds a new LDAPInterface using a schema-appropriate config
+func NewLDAPInterface(clientConfig ldaputil.LDAPClientConfig,
+	userQuery ldaputil.LDAPQueryOnAttribute,
+	groupMembershipAttributes []string,
+	userNameAttributes []string) LDAPInterface {
+	return LDAPInterface{
+		clientConfig:              clientConfig,
+		userQuery:                 userQuery,
+		userNameAttributes:        userNameAttributes,
+		groupMembershipAttributes: groupMembershipAttributes,
+		cachedUsers:               map[string]*ldap.Entry{},
+		ldapGroupToLDAPMembers:    map[string][]*ldap.Entry{},
+	}
+}
+
+var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+
+// LDAPInterface extracts the member list of an LDAP group entry from an LDAP server
+// with first-class LDAP entries for user only. The LDAPInterface is *NOT* thread-safe.
+// The LDAPInterface satisfies:
+// - LDAPMemberExtractor
+// - LDAPGroupLister
+type LDAPInterface struct {
+	// clientConfig holds LDAP connection information
+	clientConfig ldaputil.LDAPClientConfig
+
+	// userQuery holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server
+	userQuery ldaputil.LDAPQueryOnAttribute
+	// groupMembershipAttributes defines which attributes on an LDAP user entry will be interpreted as its ldapGroupUID
+	groupMembershipAttributes []string
+	// UserNameAttributes defines which attributes on an LDAP user entry will be interpreted as its name
+	userNameAttributes []string
+
+	cachePopulated bool
```

Comment:
Don't we ensure that it isn't in the [`New` method](https://github.com/openshift/origin/blob/master/pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go#L28)? If not, shouldn't we?